management/src/main/java/org/ehcache/management/registry/DefaultCapabilityBasedQuery.java

```diff
@@ -0,0 +1,302 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.management.registry;
+
+import org.ehcache.management.CallQuery;
+import org.ehcache.management.CapabilityBasedQuery;
+import org.ehcache.management.CapabilityManagement;
+import org.ehcache.management.ContextualResult;
+import org.ehcache.management.ContextualStatistics;
+import org.ehcache.management.ResultSet;
+import org.ehcache.management.StatisticQuery;
+import org.ehcache.management.providers.ManagementProvider;
+import org.terracotta.management.stats.Statistic;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author Mathieu Carbou
+ */
+final class DefaultCapabilityBasedQuery implements CapabilityBasedQuery {
+
+  private final String capabilityName;
+  private final CapabilityManagement capabilityManagement;
+
+  DefaultCapabilityBasedQuery(CapabilityManagement capabilityManagement, String capabilityName) {
+    this.capabilityManagement = capabilityManagement;
+    this.capabilityName = capabilityName;
+  }
+
+  @Override
+  public StatisticQuery.Builder queryStatistic(String statisticName) {
+    return new StatisticQueryBuilder(capabilityManagement, capabilityName, Collections.singletonList(statisticName));
+  }
+
+  @Override
+  public StatisticQuery.Builder queryStatistics(String... statisticNames) {
+    return new StatisticQueryBuilder(capabilityManagement, capabilityName, Arrays.asList(statisticNames));
+  }
+
+  @Override
+  public StatisticQuery.Builder queryStatistics(Collection<String> statisticNames) {
+    return new StatisticQueryBuilder(capabilityManagement, capabilityName, statisticNames);
+  }
+
+  @Override
+  public CallQuery.Builder call(String methodName, String[] argClassNames, Object[] args) {
+    return new CallQueryBuilder(capabilityManagement, capabilityName, methodName, argClassNames, args);
+  }
+
+  @Override
+  public CallQuery.Builder call(String methodName) {
+    return call(methodName, new String[0], new Object[0]);
+  }
+
+  static class StatisticQueryBuilder implements StatisticQuery, StatisticQuery.Builder {
```

Comment:
Just the implementation, not the interfaces. To avoid repeating code and classes. Do you see that it could cause an issue somewhere ? calls to the builder methods always create a new instance so the previous one is not change.